File: programmers/kakao#2/python/KAKAO18.py
from collections import defaultdict

def solution(gems):
    start = 0
    gem_distinct = set(gems)
    result = [0, len(gems)]
    counter = defaultdict(int)
    
    for end in range(len(gems)):
    
        counter[gems[end]] += 1
    
        while len(counter) == len(gem_distinct):
            if end - start < result[1] - result[0]:
                result = [start+1, end+1]

            counter[gems[start]] -= 1

            if counter[gems[start]] == 0:
                del counter[gems[start]]

            start += 1
    
    return result

"""
[2020 카카오 인턴] 보석 쇼핑
해결: O
"""

File: programmers/kakao#2/java/KAKAO18.java
import java.util.*;
class KAKAO18 {
    public int[] solution(String[] gems) {
        HashSet<String> gemDistinct = new HashSet<>(Arrays.asList(gems));
        Map<String, Integer> counter = new HashMap<>();
        int start = 0;
        int[] result = {0, gems.length};
        
        for(int i=0;i<gems.length;i++) {
            counter.putIfAbsent(gems[i], 0);
            counter.compute(gems[i], (key, value) -> value+1);
            
            while (gemDistinct.size() == counter.keySet().size()) {
                if (i - start < result[1] - result[0]) {
                    result[0] = start+1;
                    result[1] = i+1;
                }
                
                counter.compute(gems[start], (key, value) -> value - 1);
                
                if (counter.get(gems[start]) <= 0) {
                    counter.remove(gems[start]);
                }
                start++;
            } 
        }
        
        return result;
    }
}
/***
 * [2020 카카오 인턴] 보석 쇼핑
 * 해결: O
 */